imx: add Xiaomi zigbee gateway DGNWG05LM device

devbis opened the pull request:
Xiaomi DGNWG05LM and Aqara ZHWG11LM  are wall-mounted zigbee-to-internet gateways.

Devices specifications:
* SoC: Freescale iMX6ULL
* RAM: 256MiB
* Flash: 256MiB NAND flash
* Zigbee Chip: NXP jn5169
* Ring of RGB LEDs
* BTN0 button
* Sound chip TFA9882 + speaker


Xiaomi DGNWG05LM:
* Wireless 2.4GHz (RTL8723BS): b/g/n
* Bluetooth 4.2 integrated in RTL8723BS

Aqara ZHWG11LM:
* Wireless 2.4GHz (RTL8189ES): b/g/n (no bluetooth)

devbis committed:
bluetooth: add rtk_hciattach package with H5 support

It is required to set up bluetooth HCI on rtl8723bs chips

Signed-off-by: Ivan Belokobylskiy <belokobylskij@gmail.com>

adschm commented on target/linux/imx6/base-files/lib/upgrade/platform.sh:
```diff
@@ -60,7 +60,9 @@ platform_check_image() {
-	gw,imx6q-gw5913 )
+	gw,imx6q-gw5913 |\
+	xiaomi,dgnwg05lm |\
+	xiaomi,gateway-lumi )
```
This is only one device, so stick to the name. Same below.

devbis replied:
Left only` xiaomi,gateway-lumi` as the generic name for xiaomi/aqara devices

adschm commented on target/linux/imx6/image/lumi.mk:
```diff
@@ -0,0 +1,33 @@
+#
```
Why would we need a subtarget here?

devbis replied:
I use subtarget to set custom config for kernel to force using CONFIG_CMDLINE_FORCE with predefined partitions to use overlayfs on nand flash.

adschm replied:
So we will need a new subtarget for a single device here? Sorry, that's not acceptable for us.

devbis replied:
Is there any other way to provide a custom kernel config for a device?

adschm commented on target/linux/imx6/image/lumi.mk:
```diff
@@ -0,0 +1,33 @@
+
+
+define Device/xiaomi_dgnwg05lm
+  $(Device/xiaomi_gateway-lumi)
```
Why would we create a common node for a single device?

devbis replied:
There is a twin device from Aqara  - ZHWG11LM, it has a different rtl8189es wifi chip without integrated Bluetooth, an extra i2c chip for Homekit certificates, and Zigbee chip with an amplifier. Other hardware is identical.
I'd like to add it later but there is no 8189es driver in the kernel right now. So I made a preparation for it.

adschm replied:
Okay, fine.

adschm commented on target/linux/imx6/image/lumi.mk:
```diff
@@ -0,0 +1,33 @@
+define Device/xiaomi_gateway-lumi
+  DEVICE_VENDOR := Xiaomi
+  DEVICE_MODEL := Lumi Gateway
+  BOARD_NAME := xiaomi,gateway-lumi
```
Is BOARD_NAME still required on this target?

devbis replied:
I'd like to use the same dtb for both Xiaomi dgnwg05lm and Aqara ZHWG11LM as they are identical.
BOARD_NAME here allows me to make it equal to dtb compatible section to allow sysupgrade.

adschm replied:
That sounds like a hack that only accidentally has the desired effect (BOARD_NAME is actually an ancient remnant fixing a specific upgrade issue). Please simply create one DTS per device (using shared DTSI files), and get rid of BOARD_NAME.

devbis replied:
Rewrote this part. Left `BOARD_NAME=xiaomi,dgnwg05lm` equal to the compatible in dts to support upgrade.
It is the only way to pass the check

https://github.com/openwrt/openwrt/blob/master/package/base-files/files/lib/upgrade/nand.sh#L347-L353

adschm replied:
The trick is to use wildcards on the path, e.g. here in the generic version: https://github.com/openwrt/openwrt/blob/master/package/base-files/files/lib/upgrade/nand.sh#L278

Getting rid of BOARD_NAME is highly desirable, it's an outdated sysupgrade mechanism and quite painful if one wants to change any names.

devbis replied:
I removed BOARD_NAME and fixed upgrade image checks

adschm commented on target/linux/imx6/image/lumi.mk:
```diff
@@ -0,0 +1,33 @@
+  	kmod-sound-core kmod-button-hotplug kmod-bluetooth \
+  	bluez-daemon bluez-libs bluez-utils alsa-utils hostapd \
+  	wpa-supplicant ca-certificates kmod-gpio-button-hotplug \
+  	kmod-i2c-core kmod-iio-core nand-utils
```
Hanging indent with one tab only.

devbis replied:
Fixed extra spaces for hanging indent

adschm commented on target/linux/imx6/image/lumi.mk:
```diff
@@ -0,0 +1,33 @@
+
+define Device/xiaomi_dgnwg05lm
+  $(Device/xiaomi_gateway-lumi)
+  DEVICE_MODEL := DGNWG05LM (RTL8723BS wifi)
```
Is the WiFi relevant here (i.e. are there multiple devices with the same name, but different Wifi)?

devbis replied:
There is Aqara ZHWG11LM (RTL8189ES wifi) with the same hardware. I want to add one more device 
Aqara ZHWG11LM (RTL8189ES wifi)

Do you think I should remove the extra mentioning of the wifi chip?

adschm replied:
Mentioning the Wifi chip in the device name would only be necessary if the devices would otherwise have the exact same name (i.e could not be distinguished otherwise).

devbis replied:
Fixed this in the new patch

adschm commented on target/linux/imx6/image/lumi.mk:
```diff
@@ -0,0 +1,33 @@
+  $(Device/xiaomi_gateway-lumi)
+  DEVICE_MODEL := DGNWG05LM (RTL8723BS wifi)
+  DEVICE_PACKAGES += kmod-rtl8723bs rtl8723bs-firmware
+  SUPPORTED_DEVICES += xiaomi,gateway-lumi
```
Why add a second name here?

devbis replied:
The default supported device is "xiaomi,dgnwg05lm" which doesn't match the one in the DTB compatible section.
I want to use the same DTB for both zigbee gateways and allow sysupgrade at the same time.

I use "xiaomi,gateway-lumi" for both devices in dtb compatible.

adschm replied:
The compatible should be unique. Please create different DTS files with different compatible for different devices.

Apart from that, note the difference between "+=" and ":=".

adschm commented on target/linux/imx6/lumi/base-files/etc/uci-defaults/50_disable-removing-wdev:
```diff
@@ -0,0 +1,5 @@
+. /lib/functions.sh
+
+sed -i 's/iw dev "$wdev" del/# iw dev "$wdev" del/' /lib/netifd/wireless/mac80211.sh
```
This should be fixed properly somehow instead.

devbis replied:
Removed this patch from the PR

adschm commented:
Second and third patch lack Signed-off-by. Third patch doesn't need to have device-name twice in it, apart from that it's not device-specific anyway (and should be reworked to be specific for the actual issue ...)

devbis commented on target/linux/imx6/Makefile:
```diff
@@ -10,10 +10,15 @@ BOARDNAME:=Freescale i.MX 6
 
+define Target/Description
+	Build firmware image for Freescale i.MX 6 SoC devices.
+endef
```
Removed this change

devbis committed:
uboot: imx: add xiaomi dgnwg05lm board support

Boot from nand, squashfs or ubifs (set env bootargs=$(bootargs_ubi) or set env bootargs=$(bootargs_squash))
Added pwm leds and button (state available on env)

Contains driver update from NXP: MLK-11719-4: mtd: gpmi: change the BCH layout setting for large oob NAND

Signed-off-by: Ivan Belokobylskiy <belokobylskij@gmail.com>

devbis commented:
I reworked my changes to support the device in the generic sub-target with building u-boot for the device

adschm commented on target/linux/imx6/image/Makefile:
```diff
@@ -235,4 +235,38 @@ define Device/apalis
+define Device/xiaomi_dgnwg05lm
+  $(Device/xiaomi_gateway-lumi)
+  DEVICE_MODEL := DGNWG05LM
+  BOARD_NAME := xiaomi,dgnwg05lm
```
BOARD_NAME should not be needed anymore.

devbis replied:
As I said https://github.com/openwrt/openwrt/pull/4053#discussion_r615370393
it is the only way to pass the check for nand upgrade. Otherwise, it keeps underscore in the directory path inside sysupgrade binary instead of a comma. https://github.com/openwrt/openwrt/blob/master/package/base-files/files/lib/upgrade/nand.sh#L347-L353

adschm replied:
It's "the only way" just because you think it's impossible to replace a comma by an underscore? Please, try to implement a wildcard-based approach so we get rid of these issues...

devbis replied:
Removed BOARD_NAME in favor of more tricky `platform_check_image()`

adschm commented on target/linux/imx6/image/Makefile:
```diff
@@ -235,4 +235,38 @@ define Device/apalis
 
+define Device/xiaomi_gateway-lumi
+  DEVICE_VENDOR := Xiaomi
+  DEVICE_MODEL := Lumi Gateway
```
This is overwritten, so you can drop it here.

devbis replied:
Squashed into a single node so it is not relevant anymore

adschm commented on target/linux/imx6/image/Makefile:
```diff
@@ -235,4 +235,38 @@ define Device/apalis
+endef
+
+define Device/xiaomi_dgnwg05lm
+  $(Device/xiaomi_gateway-lumi)
```
Since you only have one device for now, I don't think we need the "common" node yet.

devbis replied:
Squashed into a single node

adschm commented on target/linux/imx6/modules.mk:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: GPL-2.0-only
+#
+# Copyright (C) 2019 OpenWrt.org
```
You are not OpenWrt, so you can't add this Copyright ...

devbis replied:
Removed.

adschm commented:
You should probably split out more of the kernel patches from the device support commit (unless they are directly related like the DTS ...).

devbis committed:
kernel: iio: add vf610 driver support

This patch adds kernel modules for Freescale VF610 ADC driver.

Signed-off-by: Ivan Belokobylskiy <belokobylskij@gmail.com>

devbis committed:
kernel: sound: add support for NXP TFA9882 DAC

Add in-kernel support for TFA9882 sound chip on
NXP devices, such as Xiaomi/Aqara Zigbee gateways

Signed-off-by: Ivan Belokobylskiy <belokobylskij@gmail.com>

devbis committed:
imx: add Xiaomi Gateway DGNWG05LM support

Xiaomi DGNWG05LM is a wall-mounted zigbee-to-internet gateway.

Device specifications:
* SoC: Freescale iMX6ULL
* RAM: 256MiB
* Flash: 256MiB NAND flash
* Wireless 2.4GHz (RTL8723BS): b/g/n
* Bluetooth 4.2 integrated in RTL8723BS
* Zigbee Chip: NXP jn5169
* Ring of RGB leds
* BTN0 button
* Sound chip TFA9882 + speaker

Flashing instructions:

Flashing requires a flashing tool from NXP: https://github.com/NXPmicro/mfgtools

* solder wires and attach to system UART and usb connector
* stop uboot by pressing any key
* enter `bmode usb` to enable flashing over USB
* run `mfgtoolcli` on the main computer to write kernel, dtb and rootfs to the device

Known issues:
- RTL8723BS driver is buggy and may cause freezes and kernel panics
- Removing wlan dev is not supported due to driver freezes.

Signed-off-by: Ivan Belokobylskiy <belokobylskij@gmail.com>

devbis commented:
> You should probably split out more of the kernel patches from the device support commit (unless they are directly related like the DTS ...).

Good catch! Replaced with a list of separate commits for each feature.

devbis committed:
kernel: wifi: realtek: 8189es and 8723bs support

Add out-of-tree drivers for SDIO wifi chips.
It contains the latest drivers available at the moment:

- 8189es: v5.8.9
- 8723bs: v5.2.17.1

Support concurrent mode and linux >= 5.8

Signed-off-by: Ivan Belokobylskiy <belokobylskij@gmail.com>

devbis committed:
imx: add Aqara Gateway ZHWG11LM support

Aqara ZHWG11LM is a wall-mounted zigbee-to-internet gateway.
It is almost a twin of Xiaomi DGNWG05LM with another wifi chip.

Device specifications:
* SoC: Freescale iMX6ULL
* RAM: 256MiB
* Flash: 256MiB NAND flash
* Wireless 2.4GHz (RTL8189ES): b/g/n
* Zigbee Chip: NXP jn5169
* Ring of RGB leds
* BTN0 button
* Sound chip TFA9882 + speaker

Flashing instructions:

Flashing requires a flashing tool from NXP: https://github.com/NXPmicro/mfgtools

* solder wires and attach to system UART and usb connector
* stop uboot by pressing any key
* enter `bmode usb` to enable flashing over USB
* run `mfgtoolcli` on the main computer to write kernel, dtb and rootfs to the device

Known issues:
- RTL8189ES driver is buggy and may cause freezes and kernel panics
Known issues:
- Removing wlan dev is not supported due to driver freezes.

Signed-off-by: Ivan Belokobylskiy <belokobylskij@gmail.com>

devbis commented:
I added external wifi drivers that support concurrent mode (allows AP and STA in several cases) to allow configuring without soldering. 

I added Aqara ZHWG11LM as well, as it has almost the same hardware

looi commented:
FYI: the "avoid kernel panic for staging rtl8723bs AP" is not needed anymore, as a fix was already merged in [kernel 5.4.126](https://cdn.kernel.org/pub/linux/kernel/v5.x/ChangeLog-5.4.126)

devbis commented:
> FYI: the "avoid kernel panic for staging rtl8723bs AP" is not needed anymore, as a fix was already merged in [kernel 5.4.126](https://cdn.kernel.org/pub/linux/kernel/v5.x/ChangeLog-5.4.126) (and other LTS kernels)

Thank you @looi ! I dropped the commit as it is not needed anymore.

adschm commented on target/linux/imx6/base-files/lib/upgrade/platform.sh:
```diff
@@ -64,7 +64,8 @@ platform_check_image() {
 		;;
-	xiaomi,dgnwg05lm )
+	xiaomi,dgnwg05lm |\
+	aqara,zhwg11lm )
```
Please ensure proper alphabetic sorting (within each "block" and for the blocks).

devbis replied:
I fixed the sorting and rebased changes on master.

devbis commented:
Fixed PKG_MIRROR_HASH for rtl8723-ol and rebased on master HEAD.

devbis commented:
Resolved conflicts after 505b7a2d08a79780f50a2adb383f57ef63bd85ac

devbis commented:
Reorganized code after d1c66eacabe11f8088f2d49e94b80dfa8a0f94f7

pepe2k commented:
@devbis are you still willing to get this merged? Personally I would prefer to get those devices supported first in upstream (kernel + U-Boot) and then backport that to OpenWrt.

devbis commented:
It could take much more time (though I can say about long time looking at the lifetime of this PR btw) and the problem is that I'm not the owner of the code. I basically adapt the code from IMX kernel tree.
I need to check WiFi drivers are built correctly against kernel 5.10. Last time I've tried there were some issues.
So, the answer is "yes", I'd like to get it merged after some checks

CGarces commented:
Hi @devbis 
Consider use https://github.com/jwrdegoede/rtl8189ES_linux as source for rtl8189ES driver.

The repository is used as mainline source of other projects like [buildroot](https://github.com/buildroot/buildroot) or [armbian](https://github.com/armbian/build) so it's more easy to share patches and fix errors between active projects. It's currently maintained and up to date with the last changes from kernel 6.1-rc

Maybe you can remove some patches like [5.15 compatibility](https://github.com/jwrdegoede/rtl8189ES_linux/commit/0eb22ce92d47c541791e1fef560916e2718084eb) or the byteorder (see https://github.com/jwrdegoede/rtl8189ES_linux/pull/86)

The build is tested against the kernels at https://kernel.ubuntu.com/~kernel-ppa/mainline/
I'm sure that works fine against kernel 5.10, but the [kernel-ppa automated build](https://github.com/jwrdegoede/rtl8189ES_linux/actions/runs/3351137071/jobs/5552389957) contains errors so the [CI build](url) is currently failing for that kernel branch.

CGarces commented:
About rtl8723bs, Bluetooth is supported by the kernel and there is a wifi driver on staging

https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/net/realtek-bluetooth.yaml
https://github.com/torvalds/linux/tree/master/drivers/staging/rtl8723bs

devbis commented:
@CGarces Thanks for the note about 8189es chip, I'm using the forked version, but I guess it can be switched to the original repo. 
As for 8723, the mainlined version doesn't support concurrent mode to allow two interfaces and for this device (without ethernet), it is the only way to configure it without UART permanently soldered. It is the main reason why using the out-of-tree driver.

Neustradamus commented:
@devbis: Any progress on this PR?

devbis committed:
kmod-bluetooth: add H5 protocol

This patch enables support for H5 (3-wire protocol)
that is used by some Realtek BT chips like rtl8723.

Signed-off-by: Ivan Belokobylskiy <belokobylskij@gmail.com>

devbis commented:
I changed repo for rtl8189 driver and added necessary patches. 
Additionalдy, since kernel 6.x built-in driver for NAND is used with `fsl,use-minimum-ecc` option enabled in DTS.
